### PR TITLE
Support oneOf, anyOf, and allOf schema combinators

### PR DESCRIPTION
**Pull request by stefanv**
Adds support for [OpenAPI v3 schema combinators](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/)

I couldn't get the tests running locally yet, but I presume they should go in `tests/test_core.py`?

### CODE REVIEW

**Comment by stefanv** [comment]
Any thoughts on this PR?

**Comment by lafrech** [comment]
Thank you for your contribution. Sorry for the delayed feedback.

This makes sense.

I'd put the tests in test_ext_marshmallow.py.

**Comment by stefanv** [comment]
@lafrech I've spent some more time on this, but I cannot figure out how to test this correctly.  I am also unsure whether there should be additional code to handle the `not` operator?  I did update the field specification from the latest version on the web.

If anyone who is more familiar with the project wants to give a hand here, they are more than welcome to push to my branch.

**Comment by killthekitten** [comment]
Hey @lafrech, I figured this needs a little push to get going. Does https://github.com/stefanv/apispec/pull/1 make sense? 

Also, I would remove the `not` definition from the original @stefanv's code and handle it in a separate pull request.

File: src/apispec/ext/marshmallow/openapi.py [hunk 1]
```diff
@@ -91,8 +91,11 @@
     "required",
     "enum",
     "type",
-    "items",
     "allOf",
+    "oneOf",
+    "anyOf",
+    "not",
```
Comment (killthekitten) [inline src/apispec/ext/marshmallow/openapi.py:97]: This is not implemented yet, @stefanv would you prefer working on it in this PR, or remove?

```suggestion
```

**Comment by stefanv** [comment]
@killthekitten Would you be interested in taking over this PR?  I see my branch is now out of date with `origin/dev` as well, and while we still desperately need this feauture, I am not familiar with the apicspec internals.

For what it's worth, we have a temporary workaround:

```
                application/json:
                  schema:
                    allOf:
                      - $ref: '#/components/schemas/SomeSchema'
                      - type: object
                        properties:
                          data:
                            type: object
                            properties:
                              ...

```

**Comment by killthekitten** [comment]
@stefanv sure, it should be fairly simple. Before I take over, I'd like to know @lafrech's point of view on this PR - is it something that is going to be merged?

**Comment by killthekitten** [comment]
@lafrech could you let me know what needs to be done to merge this? If I were to add some more changes, would it make sense to start a new PR? So far, I've added tests for `oneOf`, `anyOf` and `allOf`, and suggested to remove `not` as out of scope of the PR.

**Comment by lafrech** [comment]
I'd like to sort this and more or less related stuff (#554, #560) in 4.0.

I'm afraid you can't do much more to help. I'm just totally lacking time for this these days. Sorry about that. But it's on top of my apispec TODO list.

**Comment by killthekitten** [comment]
@lafrech thanks, I understand. Should we close this then?

**Comment by lafrech** [comment]
No, not until we're done.

**Comment by stefanv** [comment]
I know the feeling @lafreche. Hang in there, we very much appreciate the work you do!

**Comment by lafrech** [comment]
(Finally) done in #701.

**Pull request closed without merging**